File: python/chapter4/A17.py
N = int(input())
A = list(map(int, input().split()))
B = list(map(int, input().split()))

dp = [ None ] * (N+1)
dp[1] = 0
dp[2] = A[0]
for i in range(3, N+1):
    dp[i] = min(dp[i-1]+A[i-2], dp[i-2]+B[i-3])

Answer = []
Place = N
while True:
    Answer.append(Place)
    if Place == 1:
        break

    if dp[Place-1] + A[Place-2] == dp[Place]:
        Place = Place - 1
    else:
        Place = Place - 2
Answer.reverse()

Answer2 = [str(i) for i in Answer]
print(len(Answer))
print(" ".join(Answer2))

File: go/chapter4/A17/A17.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)

	scanner.Scan()
	n, _ := strconv.Atoi(scanner.Text())

	a := make([]int, n-1)
	for i := 0; i < n-1; i++ {
		scanner.Scan()
		a[i], _ = strconv.Atoi(scanner.Text())
	}

	b := make([]int, n-2)
	for i := 0; i < n-2; i++ {
		scanner.Scan()
		b[i], _ = strconv.Atoi(scanner.Text())
	}

	dp := make([]int, n)
	dp[0] = 0
	dp[1] = a[0]

	for i := 2; i < n; i++ {
		dp[i] = min(dp[i-1]+a[i-1], dp[i-2]+b[i-2])
	}

	answer := []int{}
	place := n
	for true {
		answer = append(answer, place)
		if place == 1 {
			break
		}

		if dp[place-2]+a[place-2] == dp[place-1] {
			place = place - 1
		} else {
			place = place - 2
		}
	}
	reverse(&answer)

	fmt.Println(len(answer))
	answer2 := make([]string, len(answer))
	for i := 0; i < len(answer); i++ {
		answer2 = append(answer2, strconv.Itoa(answer[i]))
	}
	fmt.Print(strings.Join(answer2, " "))
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func reverse(arr *[]int) {
	n := len(*arr)
	for i := 0; i < n/2; i++ {
		(*arr)[i], (*arr)[n-1-i] = (*arr)[n-1-i], (*arr)[i]
	}
}
